### PR TITLE
Allow sound instances to customize stream source

### PR DESCRIPTION
**Pull request by Aeronica**
This is a backport of  https://github.com/MinecraftForge/MinecraftForge/pull/8295 
I reviewed the ISound.java.patch to ensure the moj map is Sound.getPath() since there was later PR to correct the patch https://github.com/MinecraftForge/MinecraftForge/pull/8532 on the 1.18.2 branch. 

Tested using the included audio test mod.

### CODE REVIEW

File: src/test/java/net/minecraftforge/debug/client/AudioStreamTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
+/*
```
Comment (TheCurle) [inline src/test/java/net/minecraftforge/debug/client/AudioStreamTest.java]: Please rerun the license task - we use the SPDX format now.

Comment (Aeronica) [reply src/test/java/net/minecraftforge/debug/client/AudioStreamTest.java]: updated license

**Comment by Aeronica** [comment]
Looks like I need to rebase my changes on top of TheCurle's license update 4e47eab41e57460f06342c634309a16e16a72b3f  Should I try or wait for further instructions?

**Comment by TheCurle** [comment]
Yeah, you'll need to rebase & run `updateLicenses` again to regenerate the file header.

**Review by sciwhiz12 — APPROVED** [review]

**Comment by Aeronica** [comment]
Squashed the commits: "Backport #8295 Allow sound instances to customize stream source (#8595)"

**Pull request merged**